lib/appmap/middleware/remote_recording.rb

```diff
@@ -75,10 +76,50 @@ def stop_recording(req)
       end
 
       def call(env)
+        # Note: Puma config is avaliable here. For example:
+        # $ env['puma.config'].final_options[:workers]
+        # 0
+
         req = Rack::Request.new(env)
         return handle_record_request(req) if req.path == '/_appmap/record'
 
-        @app.call(env)
+        start_time = Time.now
+        # Support multi-threaded web server such as Puma by recording each thread
+        # into a separate Tracer.
+        tracer = AppMap.tracing.trace(thread: Thread.current) if record_all_requests?
+
+        @app.call(env).tap do |status, headers|
+          if tracer
+            AppMap.tracing.delete(tracer)
+
+            events = tracer.events.dup.map(&:to_h)
+
+            appmap_name = "#{req.request_method} #{req.path} (#{status}) - #{start_time.strftime('%T.%L')}"
```

Comment:
Adding request method, path and status to the metadata would also be useful for displaying this in an IDE without needing to parse `events`.

Some other ideas:
- Elapsed time of the request
- Query strings and elapsed times
- Exception details (stack trace?) if an exception failed the request

Reply:
I think there are a couple of efficient strategies we can use to implement this.

After indexing, the AppMaps list `events` first. The `http_server_request` is the first event. So, a streaming JSON parser can efficiently pick out the first event, which will be the request, and analyze it for query string, elapsed time, etc. 

The index contains `canonical.httpServerRequests` (which will contain one route in this case). The `status` code is in that file - so it's easy to quickly find out if the request succeeded or failed.

Having `exceptions.json` in the index seems like it would be a nice addition. We can handle this with an enhancement to the indexer.